### DISCUSSION TITLE
How do I get the c++20 concepts header?

### DISCUSSION THREAD

**Discussion by EricAtORS** [intel/llvm#4900] [Q&A] [answered]
I want to define my kernel using named kernel functions (functors) so that I can pass a template class.
For the template parameters, I want to use C++20 concepts to restrict and make it nicer to develop with my IDE, but it seems like it's not able to find the `<concepts>` header.

In my source folder, I can see it available in `$DPCPP_HOME/sycl_workspace/llvm/libcxx/include/concepts` but it does not get copied to the output folder of `$DPCPP_HOME/sycl_workspace/llvm/build/install/include`.

Am I missing a flag to compile my library with concepts or an option for configuring/compilng llvm?
On godbolt, it just works if I specify `-std=gnu++20`.

**Comment by alexbatashev** [comment, marked as answer]
@EricAtORS DPC++ uses your host STL implementation, not the libc++, that is part of LLVM project. That means you have to upgrade your default host compiler and its standard library (I presume, it's GCC) to a newer version.

**Reply by EricAtORS** [reply]
@alexbatashev  Thanks for the clarification.

My goal was to eventually switch our entire toolchain from gcc to clang.
Does that mean that I would need to have 2 compilers? One for regular c++ and one for sycl?

**Reply by alexbatashev** [reply]
@EricAtORS not really. Standard library is part of a C++ standard, so any compiler should provide an implementation (or means to use someone else's implementation). In Linux world the most popular implementation is GNU libstdc++. This is what most applications on your system use, and that's what we use to build libsycl.so. Now, clang tries to be compatible with gcc, which means, that it can use libstdc++ most of the time (I had some troubles in the early days of coroutines, but most of the time it works flawlessly). LLVM also provides its own standard library implementation, called libc++. While it is standard on macOS and other Apple OSes, it is not wide spread on Linux. In theory, you can build libsycl.so linked with libc++. We have experimental support for that scenario. But a) we don't test it regularly b) you may still encounter issues with other libraries, that you use. That leaves you with basically one option: to upgrade libstdc++ to the most recent version (depending on your distro, that can be done independently of gcc itself) and use it as your standard library, while using clang as a C++ compiler.
